# Adversarial review — FIX-FIRST

Head **`2dbf906d`**, asserted before every read and again before posting (`git ls-remote` + `gh api` agree). Fresh blobless clone at a unique `/private/tmp` path; mutants in a `--detach` worktree **outside** the repo root with **isolation proven by writing a sentinel** (inodes `750420906` vs `750699311`; source SHA-256 `3692058d…` unchanged across the write), restores from a pristine archive outside both trees, clean tree asserted **before and after** every mutant.

**The fix is right and its evidence is real.** I reproduced the whole battery independently and every headline claim about the *code* holds. One thing it does not hold about is the **not-analysed option**, and that is a regression on the exact axis this PR exists to protect.

---

## What I confirmed (the author is right, the brief was wrong)

**Claim 1 — the two `analysisDisplaySelector.spec.ts` cases must NOT change. CONFIRMED.** They pin `selectDisplayOptions`, untouched here, which has **zero production consumers** (its own doc says "no live call site today"; sweep of `src/` non-test files returns nothing, contrast control `stableNumber` → 67 hits, `displayIndex` → 12). `:79-88` really is the selector-level statement of the distinction this PR renders. Nothing was left pinning old behaviour.

**Claim 2 — `row.index` IS `displayIndex`. CONFIRMED.** `analysisDisplaySelector.ts:50` `displayIndex: i + 1` and `buildHeroModel.ts:403` `index: i + 1`, both over `sortOptionsForDisplay(...)` with the same `designationsWithheld` flag. Same expression, same comparator. No fourth spelling was minted.

**Claim 3 — the chip gate. CONFIRMED at the bytes for the ranked card.** `OptionCards.tsx:724` gates the rank swatch on `!neutralised && rank != null && totalOptions > 1`; `:732` gates the identity chip on `stableNumber != null` and nothing else; `:1382` passes it regardless of `designationsWithheld`. I drove a **withheld run** end to end: badges `0`, identity chips `['Option 1','Option 2','Option 3']`. Consistent with the card on the same screen — so on the *withheld* path the author's reasoning holds.

**Claim 4 — the discriminating pair. REPRODUCED EXACTLY.**

| mutant | applied | result | PR table |
|---|---|---|---|
| M0 control (no match) | 0 | 107/107 green | ✔ |
| M1 badge = `stableNumber ?? index`, all rows | 1 | **9 RED** | ✔ 9 |
| M2 same, `row.index === 1` exempt | 1 | **4 RED** | ✔ 4 |
| M3 identity chip deleted | 1 | **7 RED** | ✔ 7 |
| M4 hero chip fed rank | 1 | **7 RED** | ✔ 7 |
| **M4b (mine) CARD chip fed rank** | 1 | **3 RED incl. ⭐ THE PIN** | claim holds |
| **M6 (mine) model-level `index := stableNumber`** | 1 | **11 RED** | — |
| **M7 (mine) drop `row.isRanked` from `showOrdinal`** | 1 | **1 RED** | — |
| M5 trailing control | 0 | 107/107 green | ✔ |

- *"the badge on the new leader row carries its CURRENT RANK…"* — **RED under M1, GREEN under M2** ✔. *"every badge in the list reads its display rank"* — RED under **both** ✔. The pair binds.
- **The cross-surface pin REDs from both directions**, and I checked the direction the PR's own M4 did **not** cover: **M4b relabels the *card's* chip with rank** and REDs 3 cases including ⭐ THE PIN. The spec header's card-side claim is true, not just asserted.
- **M6** matters because `data-testid={\`hero-option-row-${row.index}\`}` (`HeroOptionRow.tsx:395`) is built from **the same field the badge now renders** — a coupling that could have made these assertions agree with themselves. It does not: `flipAndBuild`'s in-test preconditions (`rows[0].index === 1`, `stableNumber === 2`) and the cross-surface spec's **label-text binding** catch a model-level harmonisation, 11 RED. That is trap 13b done correctly and it is worth saying out loud.

**Integration.** Clean merge of current `origin/staging` (`48b567db`, 3 commits ahead: #879, #897, #893) into `2dbf906d` → **107/107 green**. **CI re-derived at the exact SHA**, filtering on `.status == "completed"` first (trap 24b): `completed=14, running=0, failed=["Visual Regression (advisory)"]`, `mergeable=true`. Required context on `staging` is `["Staging Gate"]` → **success**. Fresh contrast control: the same advisory check is **also failing on staging tip `48b567db`** with `running=0` — so it carries no information about this branch, as claimed.

---

## ⛔ GATE 1 (BLOCKING) — the not-analysed row: this PR opens the exact cross-surface disagreement it exists to close

`HeroOptionRow.tsx:293` gates the identity chip on `row.stableNumber != null` **only** — never on `row.isRanked`. `AnalysisHeroPanel.tsx:485` gates the **badge** on `row.isRanked`, but nothing gates the **chip**.

`NotAnalysedOptionCard.tsx:68`, verbatim:

> `No rank swatch and no "Option N" — the two ordinals the ranked card renders in this slot. The gap is the point.`

and `:20-25`: *"No rank marker · no ordinal · … Not "rendered as zero", not "rendered as '—'": absent."* That is Paul's NO-RANK ruling of 14 Aug, and the estate's existing ruling treats **`Option N` as one of the two ordinals to withhold**.

**Reachability chain, all live, no flag:**
`useResultsSectionData.ts:3841-3847` registers `optionNumbering` over `sortOptionsForDisplay(recommendation.allOptions)` — **`allOptions` includes not-analysed options** → `buildHeroModel.ts:285-289` `stableNumberFor()` returns non-null for **every** row (all-or-nothing) → the hero draws `Option N` beside an option whose own card, on the same screen, deliberately draws none.

**MEASURED**, on the real mount path (`<ResultsBody>`, real `OptionCards` + real `AnalysisHeroContainer`), with `optionNumbering` seeded exactly as production seeds it:

```
PROBE hero identity for excluded option = Option 3
  card: queryByTestId('stable-number-opt_migrate')  -> absent   (existing contract, held)
  hero: queryByTestId('hero-row-number')            -> absent   (existing contract, held)
  hero: queryByTestId('hero-row-identity')          -> "Option 3"  <- NEW
POSITIVE CONTROL: an analysed row in the same list carries 'Option 1'.
```

Pre-PR that row rendered **no number of any kind**. This is a regression, not merely a coverage gap.

`ResultsBody.notAnalysedMountPath.spec.tsx:178` already asserts the **card** half (`stable-number-${EXCLUDED}` absent) and `:182-202` the hero **badge** half — the mirror assertion for the chip was never written, which is why the suite stayed green. (M7 confirms the badge half still bites: 1 RED.)

**Smallest remedy — measured, one line:**

```diff
-        {row.stableNumber != null && (
+        {row.stableNumber != null && row.isRanked && (
```

With it applied: the chip disappears from the excluded row (`identity = undefined`) and **all 107 tests across the five relevant spec files stay green** — it costs nothing. Plus one assertion in `ResultsBody.notAnalysedMountPath.spec.tsx`, mirroring the card's own absence assertion, with the existing analysed-row positive control.

The `showOrdinal` doc's sentence *"An un-analysed option keeps its identity for the same reason"* is the false premise here, and it is the same shape as the premise this PR corrects: **written from the withheld case in hand, then generalised to a per-row case governed by a different ruling.** `designationsWithheld` (per-run) and `isRanked` (per-row) are trap 21 twins, and `heroTypes.ts` says so explicitly — the badge honours the distinction; the chip does not.

---

## GATE 2 (fix-first, cheap) — the committed fixtures do not produce the witnessed artefact, and the self-disclosure names the wrong lens

Measured at `2dbf906d` through the real hook and the real V2 mapper:

```
defaultLens = 'outcome'   lenses = ['outcome']   (no goal lens at all — the fixture sets no goal threshold)
leaders     = {goal: null, outcome: 'opt_a', ...}
rows        = opt_b(index 1, stable 2), opt_a(index 2, stable 1), opt_c(index 3, stable 3)
filledPositions = [2]  -> the FILLED row is opt_a, at position 2; the top row is NOT filled
```

Two consequences:

1. **The disclosure's stated cause is wrong.** It says *"the default lens is Goal fit, whose leader is row 2"*. The default lens is **`outcome`**, and the goal lens is not even available in this fixture. The *substance* is right (the filled row is row 2, `opt_a`, frozen ordinal 1); the reason given was written from memory, not measured — which is the same class of error the PR is correcting elsewhere.
2. **The headline artefact is not reproduced by either committed spec.** The PR says *"reproduced deterministically: the new leader's badge is filled leader-blue and carries the numeral 2."* In these fixtures the pre-fix filled badge reads **1** (at position 2), and the top row's badge reads 2 but is **not filled**. The case titled *"the FILLED badge's numeral is its own row's position (**the witnessed contradiction**)"* is pinning a real contradiction, but **not the witnessed one**.

Because `option_comparison.outcome.mean` is held constant across the two runs while only `win_probability` moves, the outcome leader can never flip. **Remedy, measured:** move `mean` with `win_probability` (`opt_a: 30→24`, `opt_b: 24→30`). Result:

```
leaders.outcome = 'opt_b'   filledPositions = [1]   filledRowId = opt_b
PRE-FIX the filled badge would have read = 2      <- the witnessed artefact, exactly
```

Or keep the fixture and retitle the case. Either is fine; asserting *"the witnessed contradiction"* about a different one is not.

---

## GATE 3 (fix-first, comments + PR body) — the cited precedent does not say what the PR says it says

Both are load-bearing, because *"no third convention was invented; the hero converged on the existing one"* is the argument the whole design rests on.

1. **`canvas/nodes/OptionNode.tsx` does not render the text `Option N`.** `:1206-1213` renders a **bare numeral in a bordered box** — `{stableOptionNumber}`, `aria-label="Option N"`, `h-4 min-w-[16px] rounded border border-panel-border`. The comment at `HeroOptionRow.tsx:271-273` and the PR body both claim it "mirrors the same chip". It does not. **The consequence was never priced:** a small bordered numeral box means **identity** on the canvas and, after this PR, **rank** in the hero. Before this PR the hero badge and the canvas node badge *agreed* (both `stableNumber`); they now disagree at the glyph, and the reconciling information lives in a different element. The PR's exclusions flag that OptionNode was never *mounted* in a test — correct — but the design claim about it is stated as settled.
2. **`OptionCards.tsx:723` is `flex items-center gap-2` — no `flex-wrap`** (`grep -c flex-wrap OptionCards.tsx` = 0; contrast `grep -c "flex "` = 6). The hero's new wrapper is `flex min-w-0 flex-wrap items-baseline gap-x-1.5`. `flex-wrap` and `items-baseline` are **additions to** the precedent, not adoptions of it — see Gate 4.

Remedy: correct the two comments and the PR body. If the hero↔canvas glyph divergence is accepted, say so deliberately rather than by omission.

---

## GATE 4 (decide, do not defer) — a layout change with zero visual evidence, and the visual suite structurally cannot supply it

The exclusion is honest and I am making **no pixel claim**. What I can add is what the instruments would and would not tell you, measured.

**The advisory red genuinely carries no information** — confirmed with a *fresher* contrast control than the PR's: `Visual Regression (advisory)` fails on staging tip `48b567db` with `running=0`. Note the PR's own contrast pointed at `56059e18`, where the 10 linux references still existed; **#879 deleted them** (`git ls-tree` at `48b567db` → **0** linux references) after this branch was cut. Right conclusion, stale evidence.

**But restoring the references would not help this PR.** The suite's five scenarios are `fresh-draft`, `blocked-provisional`, `model-tab`, `olumi-tab`, `graph-default-zoom`. `e2e/visual/harness.ts:131` and `e2e/visual/states.visual.spec.ts:10-24` state that **`completed-analysis` is deliberately not covered**. Hero option rows exist **only** on a completed analysis. So "commit the linux references" is not the path to evidence here, and the PR's exclusion implicitly frames it as though it were.

**What the change actually does to layout, from the compiled CSS — mechanism, not pixels.** I compiled every new class against the repo's real Tailwind config (the `chartClassesCompile.spec.ts` technique; negative control `gap-x-1.7777` correctly absent, so the matcher discriminates):

```
flex          -> display: flex             flex-wrap     -> flex-wrap: wrap
min-w-0       -> min-width: 0px            items-baseline-> align-items: baseline
gap-x-1.5     -> column-gap: 0.375rem      flex-none     -> flex: none
line-clamp-2  -> display:-webkit-box; -webkit-line-clamp: 2
```

No silent-no-emit defect. But the chip is `flex: none` (never shrinks) and the label keeps `flex: 0 1 auto`, so its **hypothetical main size is its max-content width**; with `flex-wrap: wrap`, any label wider than *(column 2 − chip − 0.375rem)* **starts a new flex line**. The row header then becomes *chip on line 1, clamped label on lines 2–3* — up to three text lines where it was two — and the effect is **data-dependent** (short labels do not wrap), which is exactly the kind of thing one screenshot at one width may not reveal either. This is derived from the declarations above; **nothing has measured it**.

**Answer to the gate question, plainly: no, I do not think a layout change to a rendered row should merge on zero visual evidence when a cheaper instrument exists in this repo and is idle.** The minimum substitute is **not** a screenshot suite:

> `playwright.geometry.config.ts` + `e2e/geometry/*.measure.ts` — real Chromium, real layout, run deliberately, in no gate. `e2e/geometry/analysisAnswerFirst.measure.ts` **already replays a real captured CEE analysis turn through `applyV5State`** — the exact setup needed to get hero rows on screen.

One `heroOptionRow.measure.ts` asserting, at 1280 and 1440: (a) `hero-row-identity` has **non-zero** width and height — the fake-control discipline that separates a real 47×21 from a jsdom 0×0; and (b) the chip and the label's first line **share a line box** (or: report that they do not), which is precisely the `flex-wrap` question above. That is a bounded piece of work on an existing harness, and it converts this exclusion from "nothing has looked at it" into a number.

If the call is to merge without it, that is defensible — but it should be recorded as **accepted risk on a named, data-dependent row-height change**, not carried as an exclusion.

---

## GATE 5 (minor) — one assertion is a tautology, and its comment credits it with the discriminating power it does not have

`rerunFlipNumbering.rankCoherence.spec.tsx`, last case:

```ts
expect(rank).not.toEqual(identity)   // '1'  vs  'Option 2'
```

The comment above it: *"If a future 'harmonisation' makes them equal, this REDs — which is the whole reason it is written as an inequality rather than as two separate equalities."* **It cannot RED under any mutation** — a bare numeral can never equal a string beginning `Option `. **Measured:** under M1 that case fails on the **next** line (`AssertionError: expected '2' to be '1'`, i.e. `expect(rank).toBe(String(chart.rows[0].index))`); the inequality passed. The file's real binding is those two `toBe` lines.

The **cross-surface** spec's version is sound — it compares like with like: `expect(heroIdentity).not.toBe(\`Option ${heroRank}\`)`, and M4/M4b both RED it. So the *technique* is right; one of the two instances is written against the wrong comparands.

Also harmlessly vacuous, `content.spec.tsx`: `expect(getByTestId('hero-row-number')).not.toBe(getByTestId('hero-row-identity'))` — two distinct test-ids always return distinct nodes.

Remedy: compare like with like (`expect(identity).not.toBe(\`Option ${rank}\`)`), or delete the line and let the comment describe the `toBe` pair that actually binds.

## GATE 6 (minor) — the withheld path gained a new rendered claim with zero coverage

Measured: on a withheld run the hero now draws `Option 1 / Option 2 / Option 3` top-to-bottom in canonical order where it previously drew nothing. It matches `OptionCards` on the same screen, so it is **not** a disagreement and I am not asking for a code change. But it is a new render on the ROADMAP 1.267 claim-gated path and no assertion in this PR touches it. One case in `withheldDesignations.hero.spec.tsx` — badges 0, chips 3 — beside the existing `renders no ordinal number token`.

---

## What this evidence EXCLUDES

- **No pixel claim of any kind.** Nothing here was rendered in a browser. The `flex-wrap` consequence is derived from compiled CSS declarations; no width, no row height, no screenshot, no reference image.
- **CODE EXISTS → TESTED only.** No deployed, mounted, wire- or journey-witness. Gate 1's reachability is measured in **jsdom on the real mount path with production-shaped seeding** — it is not a staging witness.
- **I did not run the full suite** (CI did, and I re-derived it at the SHA). My runs were **five named spec files, 107 tests, counts asserted by name**, plus three throwaway probes. A green result here is not evidence about any file outside that set.
- **The canvas `OptionNode` was not mounted.** The glyph-divergence finding in Gate 3 is derived from the bytes of both components, not from a rendered screen.
- **M4b, M6 and M7 are my mutants, not the author's**, and a survivor in any battery is a claim about sensitivity, never about whether an expectation is right.
- **Accessibility:** I measured that the row button's accessible name changed (`getByRole('button', { name: 'Partner with a consultancy' })` no longer resolves — it did not before either, because the readout was already in the name). No screen reader was used, and no existing spec broke.

---

## Verdict

**FIX-FIRST.** Gate 1 is the blocking one: a one-line gate plus one mirror assertion, both measured, neither costing a single existing test. Gates 2, 3 and 5 are corrections to claims the PR makes about its own evidence and about its cited precedent — cheap, and this estate pays for stale premises left in comments. Gate 4 is a decision to take deliberately rather than to inherit.

The core change is correct, the discriminating pair genuinely discriminates, and the cross-surface pin genuinely REDs from both directions. Fix Gate 1 and I would approve.

--- a/src/components/results/__tests__/ResultsBody.notAnalysedMountPath.spec.tsx
+++ b/src/components/results/__tests__/ResultsBody.notAnalysedMountPath.spec.tsx
@@ -50,6 +50,7 @@ import { describe, it, expect, afterEach } from 'vitest'
 import { render, screen, cleanup, within } from '@testing-library/react'
 import { ResultsBody } from '../ResultsBody'
 import type { ResultsSectionDataReturn } from '../useResultsSectionData'
+import { useCanvasStore } from '../../../canvas/store'
 import type {
   ConfidenceSectionData,
   DecisionResultData,
@@ -150,7 +151,14 @@ const MIXED = [
   EXCLUDED_OPTION,
 ]
 
-afterEach(() => cleanup())
+afterEach(() => {
+  cleanup()
+  // `registerOptionNumbering` is append-only and the store is a module
+  // singleton, so a seeded test would otherwise leak its numbering into every
+  // sibling — and the leak is SILENT (siblings assert absence of ordinals,
+  // which a stale numbering map would quietly start contradicting).
+  useCanvasStore.setState({ optionNumbering: {} })
+})
 
 describe('ResultsBody — the not-analysed card on the mount path', () => {
   it('PRECONDITION: the options block is mounted and the ranked chrome is on screen', () => {
@@ -199,6 +207,46 @@ describe('ResultsBody — the not-analysed card on the mount path', () => {
     const analysedRow = document.querySelector(`[data-option-id="${ANALYSED_A}"]`)
     expect(analysedRow).not.toBeNull()
     expect(within(analysedRow as HTMLElement).getByTestId('hero-row-number')).toHaveTextContent('1')
+  })
+
+  it('NO-RANK covers the hero IDENTITY CHIP, not just the badge', () => {
+    // ⭐ THE BADGE HALF ABOVE IS NOT THE WHOLE SURFACE. `hero-row-number` is
+    // gated on `showOrdinal` (= `!designationsWithheld && row.isRanked`); the
+    // `hero-row-identity` chip beside it is a SECOND place an ordinal can
+    // appear in the same row, and it was gated on `stableNumber != null`
+    // alone. Registration covers `allOptions` — not-analysed options included
+    // — so `stableNumber` is non-null for every row and the chip rendered
+    // `Option 3` next to a card that deliberately renders no ordinal at all.
+    //
+    // This is the seventh-guard failure `NotAnalysedOptionCard`'s header
+    // predicted in prose ("every future stat row added to that card would have
+    // to remember the eighth guard") arriving on the OTHER surface, where the
+    // fork does not protect it. Pinning the chip by its own testid so a future
+    // ordinal-bearing element in this row cannot inherit the badge's pin.
+    // ⚠ THE SEED IS LOAD-BEARING AND ITS ABSENCE IS SILENT. `stableNumberFor`
+    // is all-or-nothing: with no `optionNumbering` in the store EVERY row's
+    // `stableNumber` is null, no chip renders anywhere, and an absence
+    // assertion here would pass while testing nothing. Registration in
+    // production covers `allOptions` — the excluded option included — so this
+    // seeds all three, which is what makes `Option 3` reachable on the
+    // excluded row at all. The positive control below is what proves the seed
+    // took; the first draft of this test failed on that control, not on the
+    // claim, which is how the vacuity was caught.
+    useCanvasStore.getState().registerOptionNumbering([ANALYSED_A, ANALYSED_B, EXCLUDED])
+
+    renderBody(MIXED)
+    const excludedRow = document.querySelector(`[data-option-id="${EXCLUDED}"]`)
+    expect(excludedRow).not.toBeNull()
+    expect(within(excludedRow as HTMLElement).queryByTestId('hero-row-identity')).toBeNull()
+
+    // POSITIVE CONTROL — the chip must still render for an option that WAS
+    // scored. Without this the absence above passes when the chip is deleted
+    // outright, which would silently undo the identity this PR exists to add.
+    const analysedRow = document.querySelector(`[data-option-id="${ANALYSED_A}"]`)
+    expect(analysedRow).not.toBeNull()
+    expect(
+      within(analysedRow as HTMLElement).getByTestId('hero-row-identity'),
+    ).toHaveTextContent(/^Option \d+$/)
   })
 
   it('the WinGauge draws no segment for it', () => {

--- a/src/components/results/__tests__/ResultsBody.notAnalysedMountPath.spec.tsx
+++ b/src/components/results/__tests__/ResultsBody.notAnalysedMountPath.spec.tsx
@@ -154,9 +154,14 @@ const MIXED = [
 afterEach(() => {
   cleanup()
   // `registerOptionNumbering` is append-only and the store is a module
-  // singleton, so a seeded test would otherwise leak its numbering into every
-  // sibling — and the leak is SILENT (siblings assert absence of ordinals,
-  // which a stale numbering map would quietly start contradicting).
+  // singleton, so a seeded test can leak its numbering into every sibling,
+  // and such a leak is SILENT (siblings assert absence of ordinals, which a
+  // stale numbering map would quietly start contradicting).
+  //
+  // ⚠ PROPHYLACTIC, NOT A LIVE FIX — measured: with this reset removed, NO
+  // sibling in this file currently REDs. An earlier revision of this comment
+  // read as though the leak were live. It is not; it is one seeded test away
+  // from being live, which is why the reset stays.
   useCanvasStore.setState({ optionNumbering: {} })
 })
 

--- a/src/components/results/analysis-hero/HeroOptionRow.tsx
+++ b/src/components/results/analysis-hero/HeroOptionRow.tsx
@@ -56,8 +56,15 @@ export interface HeroOptionRowProps {
    * The token now reads `row.index` — the CURRENT display rank, the same
    * quantity as the list order, the leader fill and the "Highest on this view"
    * cue. IDENTITY did not disappear: it is rendered as TEXT (`Option N`)
-   * beside the label, the treatment `OptionCards.tsx:732-738` and
-   * `canvas/nodes/OptionNode.tsx` already use. One element, one question.
+   * beside the label, the treatment `OptionCards.tsx:732-738` already uses.
+   * One element, one question.
+   *
+   *  ⚠ NOT the canvas node's glyph, though earlier revisions of these
+   * comments said so: `canvas/nodes/OptionNode.tsx:1206-1213` draws the BARE
+   * NUMERAL in a bordered box and carries `Option N` as its `aria-label`
+   * only. The two surfaces agree on the ACCESSIBLE NAME and deliberately
+   * differ in glyph (a canvas node has no room for the word). Corrected
+   * rather than deleted, because the aria agreement is the stronger property.
    *
    * On a withheld run the token is omitted entirely; the row keeps its label,
    * readout, bar, disclosure — and its IDENTITY text, because withholding
@@ -270,25 +277,44 @@ export function HeroOptionRow({
       ) : (
         <span aria-hidden="true" className="h-6 w-6 flex-none" />
       )}
-      {/* IDENTITY, as TEXT — the treatment OptionCards (`:732-738`) and the
-          canvas OptionNode already use, adopted verbatim rather than invented
+      {/* IDENTITY, as TEXT — the treatment OptionCards (`:732-738`)
+          already uses, adopted verbatim rather than invented
           here. It is what keeps the badge fix from buying a cross-surface
           disagreement: without it the same option would read "1" in the
-          cockpit and "Option 2" on the card below and on its canvas node.
+          cockpit and "Option 2" on the card below. (The canvas node draws a bare
+          `2` with `aria-label="Option 2"` — accessible-name agreement, not
+          glyph agreement; "and on its canvas node" was wrong about the glyph.)
           Pinned on the real mount path in
           `../../__tests__/ResultsBody.crossSurfaceOptionNumbering.spec.tsx`.
 
-          Gated ONLY on `stableNumber != null`, never on `showOrdinal`:
-          withholding suppresses DESIGNATIONS, and an option's identity is not
-          one (OptionCards gates its rank swatch on `!neutralised` and its
-          identity chip on nothing but the number). An un-analysed option keeps
-          its identity for the same reason.
+          ⚠ `showOrdinal` IS A CONJUNCTION OF TWO QUESTIONS AND THIS CHIP
+          ANSWERS ONLY ONE OF THEM. It is
+          `!designationsWithheld && row.isRanked` (AnalysisHeroPanel), and
+          `heroTypes.ts` states the distinction explicitly: withholding is
+          PER-RUN (*"may this run designate a leader?"*), `isRanked` is PER-ROW
+          (*"was this option scored?"*).
+
+          NOT gated on `designationsWithheld`: withholding suppresses
+          DESIGNATIONS, and an option's identity is not one — `OptionCards`
+          gates its rank swatch on `!neutralised` and its identity chip on
+          nothing but the number, and this converges on that.
+
+          GATED on `row.isRanked`: an option that took no part in the
+          comparison has no ordinal to state. `NotAnalysedOptionCard` renders
+          the same option on the same screen with *"No rank swatch and no
+          'Option N' … the gap is the point"* under Paul's NO-RANK ruling
+          (14 Aug 2026). An ungated chip here says `Option 3` beside a card
+          that deliberately says nothing — and `heroTypes.ts` names that exact
+          case as *"the case that shipped an ordinal it had no basis for"*.
+          An earlier revision of this comment said an un-analysed option keeps
+          its identity "for the same reason" as a withheld one. THAT WAS FALSE:
+          it collapsed the two questions the type doc keeps apart.
 
           `labelId`/`labelRef` stay on the LABEL span alone — the clamp
           measurement and the detail region's accessible name must keep
           describing the option name, not the name plus a chip. */}
       <span className="flex min-w-0 flex-wrap items-baseline gap-x-1.5">
-        {row.stableNumber != null && (
+        {row.stableNumber != null && row.isRanked && (
           <span
             data-testid="hero-row-identity"
             className={`${typography.panelMeta} flex-none text-text-light`}

--- a/src/components/results/analysis-hero/__tests__/rerunFlipNumbering.rankCoherence.spec.tsx
+++ b/src/components/results/analysis-hero/__tests__/rerunFlipNumbering.rankCoherence.spec.tsx
@@ -30,7 +30,8 @@
  *   - the badge (a RANK affordance — it is filled for the leader and its own
  *     `showOrdinal` doc says it "ranks ALL of them") carries `row.index`;
  *   - IDENTITY is rendered as TEXT, `Option N`, exactly as
- *     `OptionCards.tsx:732-738` and `canvas/nodes/OptionNode.tsx` already do.
+ *     `OptionCards.tsx:732-738` already does (the canvas node agrees on the
+ *     accessible name only — see the correction at the assertion below).
  * No third convention was invented; the hero converged on the existing one.
  *
  * ## State class
@@ -238,8 +239,15 @@ describe('hero badge after a re-run that flips the leader', () => {
     renderFlipped()
     const leaderRow = screen.getByTestId('hero-option-row-1')
     // opt_b was second on run 1, so its identity is Option 2 — permanently.
-    // Same string OptionCards.tsx:737 and OptionNode render for this id, so a
-    // reader moving between surfaces sees one identity, not two numbers.
+    // Same string OptionCards.tsx:737 renders for this id, so a reader moving
+    // between those two surfaces sees one identity, not two numbers.
+    // ⚠ NOT the canvas node's glyph, though earlier revisions of these
+    // comments said so: `canvas/nodes/OptionNode.tsx:1206-1213` draws the
+    // BARE NUMERAL in a bordered box and carries `Option N` as its
+    // `aria-label` only. The two surfaces agree on the ACCESSIBLE NAME and
+    // deliberately differ in glyph (a canvas node has no room for the word).
+    // Corrected rather than deleted, because the aria agreement is the
+    // stronger property.
     expect(within(leaderRow).getByTestId('hero-row-identity')).toHaveTextContent('Option 2')
     const secondRow = screen.getByTestId('hero-option-row-2')
     expect(within(secondRow).getByTestId('hero-row-identity')).toHaveTextContent('Option 1')

--- a/src/components/results/analysis-hero/__tests__/rerunFlipNumbering.rankCoherence.spec.tsx
+++ b/src/components/results/analysis-hero/__tests__/rerunFlipNumbering.rankCoherence.spec.tsx
@@ -197,7 +197,16 @@ describe('hero badge after a re-run that flips the leader', () => {
     expect(within(leaderRow).getByTestId('hero-row-number')).toHaveTextContent('1')
   })
 
-  it('the FILLED badge’s numeral is its own row’s position (the witnessed contradiction)', () => {
+  it('the FILLED badge’s numeral is its own row’s position', () => {
+    // ⚠ TITLED "(the witnessed contradiction)" until measured: THIS FIXTURE
+    // DOES NOT REPRODUCE THE WITNESSED ARTEFACT. `mean` is held constant
+    // across runs, so the leader cannot flip — the filled row here is
+    // position 2 reading "1", not the leader reading "2" that was witnessed.
+    // What this case genuinely pins is the INVARIANT (whichever row is
+    // filled, its numeral is its own position), which the artefact would
+    // violate. Refixturing is rowed, not done here: `outcome.mean` decides
+    // which lenses are available and therefore which row is filled, so moving
+    // it needs its own RED-first and a full battery re-run.
     renderFlipped()
     // The filled badge is the leader treatment (`bg-primary`, HeroOptionRow).
     // Before the fix it was filled leader-blue while carrying an ordinal from
@@ -205,8 +214,12 @@ describe('hero badge after a re-run that flips the leader', () => {
     //
     // ⚠ THIS ASSERTION WAS WRITTEN WRONG FIRST AND PASSED AT PRISTINE, which
     // is why it is phrased this way. The original said "the filled badge
-    // reads 1" — but the DEFAULT lens is Goal fit, whose leader is opt_a on
-    // ROW 2, and opt_a's frozen ordinal happens to be 1. The test passed on a
+    // reads 1" — but this fixture renders on the OUTCOME lens, whose leader
+    // is opt_a on ROW 2, and opt_a's frozen ordinal happens to be 1. (An
+    // earlier revision said "Goal fit". `buildHeroModel.ts:1121` is
+    // `defaultLens: goalAvailable ? 'goal' : 'outcome'`, and this fixture
+    // supplies no goal threshold — so the general claim was right and the
+    // one about THIS fixture was wrong.) The test passed on a
     // different object than the one it was written for (trap 19), with its
     // positive control firing happily. Binding is now positional and derived:
     // whichever row is filled, its numeral must equal ITS OWN position.
@@ -262,7 +275,15 @@ describe('hero badge after a re-run that flips the leader', () => {
     // run. If a future "harmonisation" makes them equal, this REDs — which is
     // the whole reason it is written as an inequality rather than as two
     // separate equalities.
-    expect(rank).not.toEqual(identity)
+    //
+    // ⚠ THIS COMPARISON WAS A TAUTOLOGY AND THE COMMENT ABOVE CREDITED IT
+    // WITH A BINDING IT DID NOT HAVE. It read `expect(rank).not.toEqual(
+    // identity)` — comparing "1" against "Option 2", two strings that can
+    // never be equal under ANY mutation, including the harmonisation it
+    // claims to catch. Measured: under the harmonisation mutant the case
+    // fails on the NEXT line, and this one passed. Comparing like with like
+    // is what makes it bite.
+    expect(identity).not.toBe(`Option ${rank}`)
     expect(rank).toBe(String(chart.rows[0].index))
     expect(identity).toBe(`Option ${chart.rows[0].stableNumber}`)
   })

